## 20daysofcode challenge
**ML | DAY 7: AIR QUALITY**
- Load the data into a pandas dataframe
- Deal with missing values and incorrect data types.
- Create new features for:
- Identifying weekdays
- Months of the year.
- Morning (before 08:00),Afternoon(before 16:00) and Evening.
- Encode features where necessary(Label Encoding and or One Hot Encoding).

Dataset: https://github.com/Fortune-Adekogbe/30-Days-of-ML/tree/master/Day-7

## Import Libraries

In [1]:
import pandas as pd
import numpy as np

## Load Dataset
..into pandas dataframe

In [2]:
air_quality = pd.read_csv('AirQualityUCI.csv', sep=';')
air_quality.head()

Date      Time CO(GT)  PT08.S1(CO)  NMHC(GT) C6H6(GT)  PT08.S2(NMHC)  \
0  10/03/2004  18.00.00    2,6       1360.0     150.0     11,9         1046.0   
1  10/03/2004  19.00.00      2       1292.0     112.0      9,4          955.0   
2  10/03/2004  20.00.00    2,2       1402.0      88.0      9,0          939.0   
3  10/03/2004  21.00.00    2,2       1376.0      80.0      9,2          948.0   
4  10/03/2004  22.00.00    1,6       1272.0      51.0      6,5          836.0   

   NOx(GT)  PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)     T    RH  \
0    166.0        1056.0    113.0        1692.0       1268.0  13,6  48,9   
1    103.0        1174.0     92.0        1559.0        972.0  13,3  47,7   
2    131.0        1140.0    114.0        1555.0       1074.0  11,9  54,0   
3    172.0        1092.0    122.0        1584.0       1203.0  11,0  60,0   
4    131.0        1205.0    116.0        1490.0       1110.0  11,2  59,6   

       AH  Unnamed: 15  Unnamed: 16  
0  0,7578          NaN          NaN  
1  0,7255          NaN          NaN  
2  0,7502          NaN          NaN  
3  0,7867          NaN          NaN  
4  0,7888          NaN          NaN

### Dealing with missing values and incorrect datatypes

In [3]:
air_quality.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9471 entries, 0 to 9470
Data columns (total 17 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Date           9357 non-null   object 
 1   Time           9357 non-null   object 
 2   CO(GT)         9357 non-null   object 
 3   PT08.S1(CO)    9357 non-null   float64
 4   NMHC(GT)       9357 non-null   float64
 5   C6H6(GT)       9357 non-null   object 
 6   PT08.S2(NMHC)  9357 non-null   float64
 7   NOx(GT)        9357 non-null   float64
 8   PT08.S3(NOx)   9357 non-null   float64
 9   NO2(GT)        9357 non-null   float64
 10  PT08.S4(NO2)   9357 non-null   float64
 11  PT08.S5(O3)    9357 non-null   float64
 12  T              9357 non-null   object 
 13  RH             9357 non-null   object 
 14  AH             9357 non-null   object 
 15  Unnamed: 15    0 non-null      float64
 16  Unnamed: 16    0 non-null      float64
dtypes: float64(10), object(7)
memory usage: 1.2+ MB


- There are 9471 samples, but 9357 non-null values, meaning there are missing values in the dataset
- There is a uniform amount of missing values across all columns
- DATE and TIME column are of object datatypes, instead of date-object type
- There are two unnamed columns 'Unnamed: 15' 'Unnamed: 16' which contain no values and thus will be dropped.
- Also, its seen that columns such as 'CO(GT)','T' etc seem to have number values stored as string objects (due to the comma)

In [4]:
#drop the unnamed columns
air_quality.drop(['Unnamed: 15', 'Unnamed: 16'], axis=1, inplace=True)

In [5]:
air_quality.columns

Index(['Date', 'Time', 'CO(GT)', 'PT08.S1(CO)', 'NMHC(GT)', 'C6H6(GT)',
       'PT08.S2(NMHC)', 'NOx(GT)', 'PT08.S3(NOx)', 'NO2(GT)', 'PT08.S4(NO2)',
       'PT08.S5(O3)', 'T', 'RH', 'AH'],
      dtype='object')

#### Handling Missing Values

In [6]:
air_quality.isnull().sum()

Date             114
Time             114
CO(GT)           114
PT08.S1(CO)      114
NMHC(GT)         114
C6H6(GT)         114
PT08.S2(NMHC)    114
NOx(GT)          114
PT08.S3(NOx)     114
NO2(GT)          114
PT08.S4(NO2)     114
PT08.S5(O3)      114
T                114
RH               114
AH               114
dtype: int64

In [7]:
air_quality = air_quality.dropna()
air_quality

Date      Time CO(GT)  PT08.S1(CO)  NMHC(GT) C6H6(GT)  \
0     10/03/2004  18.00.00    2,6       1360.0     150.0     11,9   
1     10/03/2004  19.00.00      2       1292.0     112.0      9,4   
2     10/03/2004  20.00.00    2,2       1402.0      88.0      9,0   
3     10/03/2004  21.00.00    2,2       1376.0      80.0      9,2   
4     10/03/2004  22.00.00    1,6       1272.0      51.0      6,5   
...          ...       ...    ...          ...       ...      ...   
9352  04/04/2005  10.00.00    3,1       1314.0    -200.0     13,5   
9353  04/04/2005  11.00.00    2,4       1163.0    -200.0     11,4   
9354  04/04/2005  12.00.00    2,4       1142.0    -200.0     12,4   
9355  04/04/2005  13.00.00    2,1       1003.0    -200.0      9,5   
9356  04/04/2005  14.00.00    2,2       1071.0    -200.0     11,9   

      PT08.S2(NMHC)  NOx(GT)  PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  \
0            1046.0    166.0        1056.0    113.0        1692.0   
1             955.0    103.0        1174.0     92.0        1559.0   
2             939.0    131.0        1140.0    114.0        1555.0   
3             948.0    172.0        1092.0    122.0        1584.0   
4             836.0    131.0        1205.0    116.0        1490.0   
...             ...      ...           ...      ...           ...   
9352         1101.0    472.0         539.0    190.0        1374.0   
9353         1027.0    353.0         604.0    179.0        1264.0   
9354         1063.0    293.0         603.0    175.0        1241.0   
9355          961.0    235.0         702.0    156.0        1041.0   
9356         1047.0    265.0         654.0    168.0        1129.0   

      PT08.S5(O3)     T    RH      AH  
0          1268.0  13,6  48,9  0,7578  
1           972.0  13,3  47,7  0,7255  
2          1074.0  11,9  54,0  0,7502  
3          1203.0  11,0  60,0  0,7867  
4          1110.0  11,2  59,6  0,7888  
...           ...   ...   ...     ...  
9352       1729.0  21,9  29,3  0,7568  
9353       1269.0  24,3  23,7  0,7119  
9354       1092.0  26,9  18,3  0,6406  
9355        770.0  28,3  13,5  0,5139  
9356        816.0  28,5  13,1  0,5028  

[9357 rows x 15 columns]

In [8]:
air_quality.isnull().sum()

Date             0
Time             0
CO(GT)           0
PT08.S1(CO)      0
NMHC(GT)         0
C6H6(GT)         0
PT08.S2(NMHC)    0
NOx(GT)          0
PT08.S3(NOx)     0
NO2(GT)          0
PT08.S4(NO2)     0
PT08.S5(O3)      0
T                0
RH               0
AH               0
dtype: int64

#### Handling incorrect datatype

In [9]:
air_quality['Date'].head()

0    10/03/2004
1    10/03/2004
2    10/03/2004
3    10/03/2004
4    10/03/2004
Name: Date, dtype: object

In [10]:
air_quality['Time'].head()

0    18.00.00
1    19.00.00
2    20.00.00
3    21.00.00
4    22.00.00
Name: Time, dtype: object

In [11]:
air_quality['Date'] = pd.to_datetime(air_quality['Date'])
air_quality['Date'].head()

0   2004-10-03
1   2004-10-03
2   2004-10-03
3   2004-10-03
4   2004-10-03
Name: Date, dtype: datetime64[ns]

In [12]:
air_quality['Time'] = pd.to_datetime(air_quality['Time'], format='%H.%M.%S')
air_quality['Time'].head()

0   1900-01-01 18:00:00
1   1900-01-01 19:00:00
2   1900-01-01 20:00:00
3   1900-01-01 21:00:00
4   1900-01-01 22:00:00
Name: Time, dtype: datetime64[ns]

In [13]:
air_quality['Weekday'] = air_quality['Date'].dt.weekday
air_quality['Month'] = air_quality['Date'].dt.month

In [14]:
air_quality.sample(3)

Date                Time CO(GT)  PT08.S1(CO)  NMHC(GT) C6H6(GT)  \
5152 2004-11-10 1900-01-01 10:00:00    3,4       1313.0    -200.0     17,1   
194  2004-03-18 1900-01-01 20:00:00    5,1       1667.0    -200.0     26,0   
5741 2004-04-11 1900-01-01 23:00:00    2,4       1181.0    -200.0     13,3   

      PT08.S2(NMHC)  NOx(GT)  PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  \
5152         1213.0    536.0         564.0     88.0        1789.0   
194          1453.0    276.0         684.0    176.0        2051.0   
5741         1094.0    345.0         600.0     74.0        1599.0   

      PT08.S5(O3)     T    RH      AH  Weekday  Month  
5152       1359.0  19,2  73,4  1,6192        2     11  
194        1569.0  18,6  36,2  0,7676        3      3  
5741       1180.0  18,8  70,6  1,5209        6      4

###### Convert 'CO(GT)' and similar columns to float data type

In [15]:
air_quality['CO(GT)'].head()

0    2,6
1      2
2    2,2
3    2,2
4    1,6
Name: CO(GT), dtype: object

In [16]:
air_quality['CO(GT)'] = air_quality['CO(GT)'].str.replace(',', '.').astype(float)
air_quality['C6H6(GT)'] = air_quality['C6H6(GT)'].str.replace(',', '.').astype(float)
air_quality['T'] = air_quality['T'].str.replace(',', '.').astype(float)
air_quality['RH'] = air_quality['RH'].str.replace(',', '.').astype(float)
air_quality['AH'] = air_quality['AH'].str.replace(',', '.').astype(float)

In [17]:
air_quality.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9357 entries, 0 to 9356
Data columns (total 17 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   Date           9357 non-null   datetime64[ns]
 1   Time           9357 non-null   datetime64[ns]
 2   CO(GT)         9357 non-null   float64       
 3   PT08.S1(CO)    9357 non-null   float64       
 4   NMHC(GT)       9357 non-null   float64       
 5   C6H6(GT)       9357 non-null   float64       
 6   PT08.S2(NMHC)  9357 non-null   float64       
 7   NOx(GT)        9357 non-null   float64       
 8   PT08.S3(NOx)   9357 non-null   float64       
 9   NO2(GT)        9357 non-null   float64       
 10  PT08.S4(NO2)   9357 non-null   float64       
 11  PT08.S5(O3)    9357 non-null   float64       
 12  T              9357 non-null   float64       
 13  RH             9357 non-null   float64       
 14  AH             9357 non-null   float64       
 15  Weekday        9357 n

In [18]:
air_quality.head()

Date                Time  CO(GT)  PT08.S1(CO)  NMHC(GT)  C6H6(GT)  \
0 2004-10-03 1900-01-01 18:00:00     2.6       1360.0     150.0      11.9   
1 2004-10-03 1900-01-01 19:00:00     2.0       1292.0     112.0       9.4   
2 2004-10-03 1900-01-01 20:00:00     2.2       1402.0      88.0       9.0   
3 2004-10-03 1900-01-01 21:00:00     2.2       1376.0      80.0       9.2   
4 2004-10-03 1900-01-01 22:00:00     1.6       1272.0      51.0       6.5   

   PT08.S2(NMHC)  NOx(GT)  PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)  \
0         1046.0    166.0        1056.0    113.0        1692.0       1268.0   
1          955.0    103.0        1174.0     92.0        1559.0        972.0   
2          939.0    131.0        1140.0    114.0        1555.0       1074.0   
3          948.0    172.0        1092.0    122.0        1584.0       1203.0   
4          836.0    131.0        1205.0    116.0        1490.0       1110.0   

      T    RH      AH  Weekday  Month  
0  13.6  48.9  0.7578        6     10  
1  13.3  47.7  0.7255        6     10  
2  11.9  54.0  0.7502        6     10  
3  11.0  60.0  0.7867        6     10  
4  11.2  59.6  0.7888        6     10

###### Create hour column from Time column so we can divide the hours in daytime(i.e Morning, Afternoon and Evening)

In [19]:
air_quality['Hour'] = air_quality['Time'].dt.hour

In [20]:
air_quality['Hour'].astype(int)

0       18
1       19
2       20
3       21
4       22
        ..
9352    10
9353    11
9354    12
9355    13
9356    14
Name: Hour, Length: 9357, dtype: int32

In [21]:
def daytime(hour):
    """function to segment hour time to Morning, Afternoon and Evening"""
    if 0 <= hour < 8:
        return 'Morning'
    elif 8 <= hour <  16: 
        return 'Afternoon'
    else:
        return 'Evening'
    
air_quality['daytime'] = air_quality['Hour'].apply(daytime)

In [22]:
air_quality.head()

Date                Time  CO(GT)  PT08.S1(CO)  NMHC(GT)  C6H6(GT)  \
0 2004-10-03 1900-01-01 18:00:00     2.6       1360.0     150.0      11.9   
1 2004-10-03 1900-01-01 19:00:00     2.0       1292.0     112.0       9.4   
2 2004-10-03 1900-01-01 20:00:00     2.2       1402.0      88.0       9.0   
3 2004-10-03 1900-01-01 21:00:00     2.2       1376.0      80.0       9.2   
4 2004-10-03 1900-01-01 22:00:00     1.6       1272.0      51.0       6.5   

   PT08.S2(NMHC)  NOx(GT)  PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)  \
0         1046.0    166.0        1056.0    113.0        1692.0       1268.0   
1          955.0    103.0        1174.0     92.0        1559.0        972.0   
2          939.0    131.0        1140.0    114.0        1555.0       1074.0   
3          948.0    172.0        1092.0    122.0        1584.0       1203.0   
4          836.0    131.0        1205.0    116.0        1490.0       1110.0   

      T    RH      AH  Weekday  Month  Hour  daytime  
0  13.6  48.9  0.7578        6     10    18  Evening  
1  13.3  47.7  0.7255        6     10    19  Evening  
2  11.9  54.0  0.7502        6     10    20  Evening  
3  11.0  60.0  0.7867        6     10    21  Evening  
4  11.2  59.6  0.7888        6     10    22  Evening

#### Label Encoding

In [23]:
#encode daytime column
air_quality = pd.get_dummies(air_quality, columns=['daytime'])
air_quality.head()

Date                Time  CO(GT)  PT08.S1(CO)  NMHC(GT)  C6H6(GT)  \
0 2004-10-03 1900-01-01 18:00:00     2.6       1360.0     150.0      11.9   
1 2004-10-03 1900-01-01 19:00:00     2.0       1292.0     112.0       9.4   
2 2004-10-03 1900-01-01 20:00:00     2.2       1402.0      88.0       9.0   
3 2004-10-03 1900-01-01 21:00:00     2.2       1376.0      80.0       9.2   
4 2004-10-03 1900-01-01 22:00:00     1.6       1272.0      51.0       6.5   

   PT08.S2(NMHC)  NOx(GT)  PT08.S3(NOx)  NO2(GT)  ...  PT08.S5(O3)     T  \
0         1046.0    166.0        1056.0    113.0  ...       1268.0  13.6   
1          955.0    103.0        1174.0     92.0  ...        972.0  13.3   
2          939.0    131.0        1140.0    114.0  ...       1074.0  11.9   
3          948.0    172.0        1092.0    122.0  ...       1203.0  11.0   
4          836.0    131.0        1205.0    116.0  ...       1110.0  11.2   

     RH      AH  Weekday  Month  Hour  daytime_Afternoon  daytime_Evening  \
0  48.9  0.7578        6     10    18                  0                1   
1  47.7  0.7255        6     10    19                  0                1   
2  54.0  0.7502        6     10    20                  0                1   
3  60.0  0.7867        6     10    21                  0                1   
4  59.6  0.7888        6     10    22                  0                1   

   daytime_Morning  
0                0  
1                0  
2                0  
3                0  
4                0  

[5 rows x 21 columns]

In [24]:
air_quality.head()

Date                Time  CO(GT)  PT08.S1(CO)  NMHC(GT)  C6H6(GT)  \
0 2004-10-03 1900-01-01 18:00:00     2.6       1360.0     150.0      11.9   
1 2004-10-03 1900-01-01 19:00:00     2.0       1292.0     112.0       9.4   
2 2004-10-03 1900-01-01 20:00:00     2.2       1402.0      88.0       9.0   
3 2004-10-03 1900-01-01 21:00:00     2.2       1376.0      80.0       9.2   
4 2004-10-03 1900-01-01 22:00:00     1.6       1272.0      51.0       6.5   

   PT08.S2(NMHC)  NOx(GT)  PT08.S3(NOx)  NO2(GT)  ...  PT08.S5(O3)     T  \
0         1046.0    166.0        1056.0    113.0  ...       1268.0  13.6   
1          955.0    103.0        1174.0     92.0  ...        972.0  13.3   
2          939.0    131.0        1140.0    114.0  ...       1074.0  11.9   
3          948.0    172.0        1092.0    122.0  ...       1203.0  11.0   
4          836.0    131.0        1205.0    116.0  ...       1110.0  11.2   

     RH      AH  Weekday  Month  Hour  daytime_Afternoon  daytime_Evening  \
0  48.9  0.7578        6     10    18                  0                1   
1  47.7  0.7255        6     10    19                  0                1   
2  54.0  0.7502        6     10    20                  0                1   
3  60.0  0.7867        6     10    21                  0                1   
4  59.6  0.7888        6     10    22                  0                1   

   daytime_Morning  
0                0  
1                0  
2                0  
3                0  
4                0  

[5 rows x 21 columns]